### PR TITLE
feat(es/lint): implement valid-typeof rule

### PR DESCRIPTION
**Pull request by ArturAralin**
**Description:**

Implement valid-typeof rule

**BREAKING CHANGE:**

No

### CODE REVIEW

File: crates/swc_ecma_lints/src/rules/valid_typeof.rs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+impl Visit for ValidTypeof {
+    noop_visit_type!();
+
+    fn visit_bin_expr(&mut self, bin_expr: &BinExpr) {
+        if let op!("==") | op!("===") | op!("!=") | op!("!==") = bin_expr.op {
+            match (bin_expr.left.as_ref(), bin_expr.right.as_ref()) {
+                // case typeof x === "type"
+                (
+                    Expr::Unary(UnaryExpr {
+                        op: UnaryOp::TypeOf,
```
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/valid_typeof.rs:81]: Can you change this to `op: op!("typeof")`?

File: crates/swc_ecma_lints/src/rules/valid_typeof.rs [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+                    }),
+                    Expr::Lit(Lit::Str(Str { value, .. })),
+                ) => {
+                    self.check(bin_expr.span, &*value);
+                }
+                // case "type" === typeof x
+                (
+                    Expr::Lit(Lit::Str(Str { value, .. })),
+                    Expr::Unary(UnaryExpr {
+                        op: UnaryOp::TypeOf,
```
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/valid_typeof.rs:92]: Same here

File: crates/swc_ecma_lints/src/rules/valid_typeof.rs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                        ..
+                    }),
+                ) => {
+                    self.check(bin_expr.span, &*value);
+                }
+                // case typeof x === typeof y
+                (
+                    Expr::Unary(UnaryExpr {
+                        op: UnaryOp::TypeOf,
```
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/valid_typeof.rs:101]: Same here

File: crates/swc_ecma_lints/src/rules/valid_typeof.rs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+    fn check(&self, span: Span, str_operand: &str) {
+        if !VALID_TYPES.contains(&str_operand) {
+            self.emit_report(span);
+        }
+    }
+}
+
+impl Visit for ValidTypeof {
+    noop_visit_type!();
```
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/valid_typeof.rs:73]: Can you remove this line?

File: crates/swc_ecma_lints/src/rules/valid_typeof.rs [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+                    self.check(bin_expr.span, &*value);
+                }
+                // case typeof x === typeof y
+                (
+                    Expr::Unary(UnaryExpr {
+                        op: UnaryOp::TypeOf,
+                        ..
+                    }),
+                    Expr::Unary(UnaryExpr {
+                        op: UnaryOp::TypeOf,
```
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/valid_typeof.rs:105]: Same here, `op!` supports all types of operations.

File: crates/swc_ecma_lints/src/rules/valid_typeof.rs [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+                        ..
+                    }),
+                ) => {}
+                // case typeof x === foo()
+                (
+                    Expr::Unary(UnaryExpr {
+                        op: UnaryOp::TypeOf,
```
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/valid_typeof.rs:112]: Same here

File: crates/swc_ecma_lints/src/rules/valid_typeof.rs [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+                ) => {
+                    if self.require_string_literals {
+                        self.emit_report(bin_expr.span);
+                    }
+                }
+                // case foo() === typeof x
+                (
+                    _,
+                    Expr::Unary(UnaryExpr {
+                        op: UnaryOp::TypeOf,
```
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/valid_typeof.rs:125]: Same here

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thank you!

File: crates/swc_ecma_lints/src/rules/valid_typeof.rs (hunk 1 above)
Comment (ArturAralin) [reply crates/swc_ecma_lints/src/rules/valid_typeof.rs:81]: Forget about it every time 😅

**Review by kdy1 — APPROVED** [review]
Thank you!

---

swc-bump:
 - swc_ecma_lints

**Review by kdy1 — APPROVED** [review]

**Pull request merged**